src/WebJobs.Script.WebHost/Controllers/AdminController.cs

```diff
@@ -148,6 +148,37 @@ public IHttpActionResult Ping()
         }
 
         [HttpPost]
+        [Route("admin/host/log")]
+        [AllowAnonymous]
+        public IHttpActionResult Log(IEnumerable<HostLogEntry> logEntries)
+        {
+            var authorizationLevel = Request.GetAuthorizationLevel();
+            if (Request.IsAuthDisabled() ||
+                authorizationLevel == AuthorizationLevel.Admin ||
+                Request.IsAntaresInternalRequest())
+            {
+                foreach (var logEntry in logEntries)
+                {
+                    var traceEvent = new TraceEvent(logEntry.Level, logEntry.Message, logEntry.Source);
+                    if (!string.IsNullOrEmpty(logEntry.FunctionName))
+                    {
+                        traceEvent.Properties.Add(ScriptConstants.TracePropertyFunctionNameKey, logEntry.FunctionName);
+                    }
+                    _traceWriter.Trace(traceEvent);
+
+                    var logLevel = Utility.ToLogLevel(traceEvent.Level);
+                    _logger.Log(logLevel, new EventId(0), logEntry, null, null);
```

Comment:
I believe the best way to log this structured would be like:

```
// Note: LogLevel is already auto-logged.
// Pass the structured data -- anything that implements IEnumerable<KeyValuePair<string, object>>:
var logData = new Dictionary<string, object>
{
    [Source] = logEntry.Source,
    [FunctionName] = logEntry.FunctionName
};

// The formatter is telling you what string to treat as the 'output' for loggers that care (like Console logger). 
// In this case -- the Message
_logger.Log(logLevel, 0, logData, null, (s, e) => logEntry.Message);
```

Some background -- the object that you log needs to implement `IEnumerable<KeyValuePair<string, object>>` in order for it to log structured data. This happens automatically if you do something like `_logger.LogInformation("message {a}", valueOfA)`, but if you're using `_logger.Log()`, you've got to build it yourself.

If the data is in the string, the easiest thing is to use `FormattedLogValues` to pass the state to `Log()`. For example, here's how `LogInformation()` works: https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/LoggerExtensions.cs#L225

But in this case, the function name and source aren't there, so passing your own dictionary and message separately should work.